src/v/kafka/server/handlers/produce.cc

```diff
@@ -462,6 +468,11 @@ produce_handler::handle(request_context ctx, ss::smp_service_group ssg) {
     produce_request request;
     request.decode(ctx.reader(), ctx.header().version);
 
+    if (ctx.metadata_cache().should_reject_writes()) {
```

Comment:
i think this could just as easily be `ctx.health_monitor().should_reject_writes()` and avoid the extra complexity added into the metadata cache for receiving up-calls and polling refreshes on the health monitor.

Reply:
Health monitor only polls on the ring0 leader. Look at it's tick() codepath.

So either way we need a "new" timer. The latest patch (just force-pushed based on friday's code review comments) eliminates the callback, but still has the "always tick on shard zero" logic in metadata_cache. We could move it down into health_monitor, if we care, of course.

Reply:
I think it would make sense to keep it in the health monitor.

Reply:
Cool with me, let's see what others like @mmaslankaprv say--I was following his suggestion here.

Reply:
I guess the problem that @ajfabbri is trying to solve is not to wait for health monitor update. Indeed in health monitor updates are made 'on demand' i.e. only if somebody calls a function and cached metadata are older than a configurable threshold.

Reply:
We may probably be a little bit less restrictive with the threshold check and do something like:

```c++
auto f = health_monitor.get_report(...);
if (f.available()){
... validate
} else {
(void)f; // wait in backgroud. 
}

```